src/main/drivers/serial.c

```diff
@@ -37,13 +37,16 @@ uint32_t serialGetBaudRate(serialPort_t *instance)
 
 void serialWrite(serialPort_t *instance, uint8_t ch)
 {
-    instance->vTable->serialWrite(instance, ch);
+    while (!serialTxBytesFree(instance)) {
+    };
+
+	instance->vTable->serialWrite(instance, ch);
 }
```

Comment:
@blckmn : My code did provide continuous chunks of memory (tx data or rx space) with commit mechanism. It is very similar to UART DMA code, so this interface may be eventually shared, but I didn't invest enough time into it.

Also circular buffer management is quite similar for all serial ports - unifying it completely will be huge speed gain (for current interface implementation/use), it can avoid vTable indirection for buffer access (read/write single char)